csrc/preseg_passes/move_gather.cpp

```diff
@@ -0,0 +1,204 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_gather.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/indexing.h>
+#include <ops/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+bool isUnaryPointwiseOp(const Expr* expr) {
+  // Check if the expression is a UnaryOp and a pointwise operation
+  return expr->isA<UnaryOp>() && ir_utils::isPointwiseTvOp(expr);
+}
+
+bool hasGatherOp(const Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  return ir_utils::filterByType<GatherOp>(exprs).size() > 0;
+}
+
+// For now we allow unary pointwise ops like cast/neg or a squeeze.
+// We only consider squeeze which has a single dim squeezed.
+std::optional<Expr*> getAllowedLookupTvDef(const GatherOp* gather_op) {
+  if (gather_op->lookupTv()->isFusionInput()) {
+    return std::nullopt;
+  }
+
+  auto producing_expr = gather_op->lookupTv()->definition();
+  NVF_ERROR(
+      producing_expr != nullptr, "Def for Lookup tensor view is nullptr.");
+
+  if (isUnaryPointwiseOp(producing_expr)) {
+    return producing_expr;
+  }
+
+  // We handle squeeze op where only one dim is squeezed.
+  if (producing_expr->isA<SqueezeOp>()) {
+    auto squeeze_op = producing_expr->as<SqueezeOp>();
+    if (std::count(
+            squeeze_op->getSqueezeDimFlags().begin(),
+            squeeze_op->getSqueezeDimFlags().end(),
+            true) == 1) {
+      return producing_expr;
+    }
+  }
+
+  return std::nullopt;
+}
+
+auto positionOfSqueezedDim(const std::vector<bool>& squeeze_dim_flags) {
+  auto it = std::find(squeeze_dim_flags.begin(), squeeze_dim_flags.end(), true);
+  return it != squeeze_dim_flags.end()
+      ? std::distance(squeeze_dim_flags.begin(), it)
+      : -1;
+}
+
+// If the def of lookUpTv is squeeze, and we're moving the squeeze to
+// after the gather/take_along_axis, then we need to add an unsqueeze
+// to the IndexTv
+auto prepareIndexTv(Fusion* fusion, GatherOp* gather_op, Expr* def) {
+  if (def->isA<SqueezeOp>()) {
+    auto squeeze_dim_flags = def->as<SqueezeOp>()->getSqueezeDimFlags();
+    auto dim = positionOfSqueezedDim(squeeze_dim_flags);
+    return unsqueeze(gather_op->indexTv(), dim);
+  }
+  return gather_op->indexTv();
+}
+
+auto addPostGatherSqueeze(Fusion* fusion, GatherOp* new_gather, Expr* def) {
+  auto new_squeeze_output = squeeze(
+      new_gather->output(0)->as<TensorView>(),
+      def->as<SqueezeOp>()->getSqueezeDimFlags());
+  return new_squeeze_output;
```

Comment:
This would have the same problem with discarding transformation on the original squeeze output.

Reply:
To clarify, this problem will show up when the user has manually scheduled the fusion?
In the case of automatic scheduling, the pre-seg pass should run before any scheduling?

Reply:
What would be the correct way to copy these transformations?
I am assuming something like this has been done before in the nvFuser code base.

If no I guess I have to replay those transforms on the new output